contracts/chain-adapters/Arbitrum_Forwarder.sol

```diff
@@ -0,0 +1,52 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity ^0.8.0;
+
+import { ForwarderBase } from "./ForwarderBase.sol";
+import { ArbitrumL1ERC20GatewayLike } from "../interfaces/ArbitrumBridgeInterfaces.sol";
+import { AddressUtils } from "../libraries/AddressUtils.sol";
+
+/**
+ * @title Arbitrum_Forwarder
+ * @notice This contract expects to receive messages and tokens from an authorized contract on a previous layer and forwards them to
+ * contracts on the subsequent layer. It rejects messages which do not originate from a cross domain admin.
+ * @custom:security-contact bugs@across.to
+ */
+contract Arbitrum_Forwarder is ForwarderBase {
+    ArbitrumL1ERC20GatewayLike public immutable arbitrumGatewayRouter;
+
+    modifier onlyFromCrossDomainAdmin() {
+        require(msg.sender == AddressUtils._applyL1ToL2Alias(crossDomainAdmin), "ONLY_COUNTERPART_GATEWAY");
+        _;
+    }
+
+    /**
+     * @notice Constructs an Arbitrum-specific forwarder contract.
+     * @dev Since this is a proxy contract, we only set immutable variables in the constructor, and leave everything else to be initialized.
+     * This includes variables like the cross domain admin.
+     */
+    constructor(ArbitrumL1ERC20GatewayLike _arbitrumGatewayRouter) ForwarderBase() {
+        arbitrumGatewayRouter = _arbitrumGatewayRouter;
+    }
+
+    /**
+     * @notice Relays `currentLayerToken` to the next layer between the current layer and the one which contains `target`.
+     * @param currentLayerToken The current layer's address of the token to send to the subsequent layer.
+     * @param amount The amount of the token to send to the subsequent layer.
+     * @param target The contract which will ultimately receive `amount` of `currentLayerToken`.
+     * @dev The first field is discarded since it contains the address of the previous layer's token, which was used to derive this layer's token
+     * address.
+     */
+    function relayTokens(
+        address,
+        address currentLayerToken,
+        uint256 amount,
+        address target
+    ) external payable override onlyAdmin {
+        address remoteToken = remoteTokens[target][currentLayerToken] == address(0)
+            ? arbitrumGatewayRouter.calculateL2TokenAddress(currentLayerToken)
```

Comment:
The gateway router should return the correct address if the token in question is actually routed through an Arbitrum gateway. The most notable exception we know of now is USDC and CCTP, which is an example of when we would instead need to set `remoteTokens`. I think if we go with this, we will need to be careful and ensure that the remote token we are attempting to bridge to is indeed created through a `bridgeMint` call.

Reply:
how would cctp work? cctp from L1 to L2, and then canonical bridge from there? can we support that?

Reply:
For reference, there was this comment https://github.com/across-protocol/contracts/pull/610#discussion_r1765177183 on the previous PR which inherited the adapter code, and the idea was that there may be a CCTP bridge from L2-L3 which bridges native to native. This should work fine if we use CCTP for L1 -> L2, have a native on L2, and then bridge native on L2 -> bridged on L3 via the canonical bridge, since that uses an Arbitrum gateway. TLDR CCTP is just speculation. It's mainly an example if there was another non-arbitrum bridge from L2-L3, and also `calculateL2TokenAddress` should return the remote token which is minted on L2 given the input L1 token.